src/test/scala/shine/host.scala

```diff
@@ -0,0 +1,132 @@
+
+  int32_t* in = hostBufferSync(ctx, input, N * sizeof(int32_t), HOST_WRITE);
+  for (int i = 0; i < N; i++) {
+    in[i] = 0;
+  }
+
+  foo(ctx, output, N, input);
+
+  int32_t* out = hostBufferSync(ctx, output, N * sizeof(int32_t), HOST_READ);
+
+  for (int i = 0; i < N; i++) {
+    if (out[i] != 3) {
+      fprintf(stderr, "wrong output: %i\n", out[i]);
+      exit(EXIT_FAILURE);
+    }
+  }
+
+  destroyBuffer(ctx, input);
+  destroyBuffer(ctx, output);
+  destroyContext(ctx);
+  return EXIT_SUCCESS;
+}
+"""
+
+  private def checkOutput(m: shine.OpenCL.Module): Unit = {
+    val prog = shine.OpenCL.Module.translateToString(m) + main
+    util.ExecuteOpenCL(prog, "zero_copy")
+    util.ExecuteOpenCL(prog, "one_copy")
+  }
+
+  private def findCount(c: Int, r: Regex, in: String): Unit =
+    r.findAllIn(in).length shouldBe c
+
+  private def findDeviceBufferSyncWrite(c: Int, in: String): Unit =
+    findCount(c, """deviceBufferSync\(.*, DEVICE_WRITE\)""".r, in)
+
+  private def findDeviceBufferSyncRead(c: Int, in: String): Unit =
+    findCount(c, """deviceBufferSync\(.*, DEVICE_READ\)""".r, in)
+
+  test("no kernel call") {
+    val e = depFun((n: Nat) => fun((n`.`i32) ->: (n`.`i32))(in =>
+      mapSeq(add(li32(3)))(in)
+    ))
+    val m = gen.opencl.hosted.fromExpr(e)
+    val hostCode = gen.c.function.asString(m.host)
+    // println(hostCode)
+    findCount(1, """hostBufferSync\(.*, HOST_WRITE\)""".r, hostCode)
+    findCount(1, """hostBufferSync\(.*, HOST_READ\)""".r, hostCode)
+    checkOutput(m)
+  }
+
+  test("basic kernel call with fixed size") {
+    val n: Nat = 128
+    val e = fun((n`.`i32) ->: (n`.`i32))(in =>
+      oclRun(LocalSize(2), GlobalSize(32))(mapGlobal(add(li32(3)))(in))
+    )
+    val m = gen.opencl.hosted.fromExpr(e)
+    val hostCode = gen.c.function.asString(m.host)
+    // println(hostCode)
+    findDeviceBufferSyncWrite(1, hostCode)
+    findDeviceBufferSyncRead(1, hostCode)
+    // m.kernels.foreach(km => println(gen.opencl.kernel.asString(km)))
+  }
+
+  test("basic kernel call with variable size and post-process") {
+    val e = depFun((n: Nat) => fun((n`.`i32) ->: (n`.`i32))(in =>
+      oclRun(LocalSize(2), GlobalSize(n/2))(mapGlobal(add(li32(1)))(in)) |> toMem |>
+      mapSeq(add(li32(2)))
+    ))
+    val m = gen.opencl.hosted.fromExpr(e)
+    val hostCode = gen.c.function.asString(m.host)
+    // println(hostCode)
+    findCount(1, """createBuffer\(.*, HOST_READ \| DEVICE_WRITE\)""".r, hostCode)
+    findDeviceBufferSyncWrite(1, hostCode)
+    findDeviceBufferSyncRead(1, hostCode)
+    findCount(1, """hostBufferSync\(.*, HOST_WRITE\)""".r, hostCode)
+    findCount(1, """hostBufferSync\(.*, HOST_READ\)""".r, hostCode)
+    checkOutput(m)
+  }
+
+  test("two kernel calls") {
+    val e = depFun((n: Nat) => fun((n`.`i32) ->: (n`.`i32))(in =>
+      oclRun(LocalSize(2), GlobalSize(n/2))(mapGlobal(add(li32(1)))(in)) |> store(x =>
```

Comment:
Do I understand correctly that every `oclRun` has to be followed by a `toMem` /  `store` when there is further computation following (i.e., it is not the last computation in the entrie RISE program)?

Reply:
Yes, `oclRun` is always `write` and does not allocate memory, `toMem` allocates memory and goes from `write` to `read`.

Reply:
Thanks for the confirmation.